docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1766 @@
+  attribute access.
+- Fixed field types should only be used if all fields are known. An example where this
+  is not the case are collections of benchmark problems, where the set of fields depends
+  on the selected benchmark sets and other things. In such situations, dictionaries that
+  map strings to BenchmarkProblem objects are a good idea.
+- For backwards compatibility and compatibility with SciPy, we allow things we don't
+  find ideal (e.g. selecting algorithms via strings). However, the documentation should
+  mostly show our prefered way of doing things. Alternatives can be hidden in tabs and
+  expandable boxes.
+- Whenever possible, use immutable types. Whenever things need to be changeable,
+  consider using an immutable type with copy constructors for modified instances.
+  Example: instances of `Algorithm` are immutable but using `Algorithm.with_option`
+  users can create modified copies.
+- The main entry point to estimagic are functions, objects are mostly used for
+  configuration and return types. This takes the best of both worlds: we get the safety
+  and static analysis that (in Python) can only be achieved using objects but the
+  beginner friendliness and freedom provided by functions. Example: Having a `minimize`
+  function, it is very easy to add the possibility of running minimizations with
+  multiple algorithms in parallel and returning the best value. Having a `.solve` method
+  on an algorithm object would require a whole new interface for this.
+
+## Changes for optimization
+
+The following changes apply to all functions that are directly related to optimization,
+i.e. `maximize`, `minimize`, `slice_plot`, `criterion_plot`, `params_plot`,
+`count_free_params`, `check_constraints` and `OptimizeResult`.
+
+### The objective function
+
+#### Current situation
+
+The objective or criterion function is the function being optimized. Currently, it is
+called `criterion` in estimagic.
+
+The same criterion function can work for scalar, least-squares and likelihood
+optimizers. Moreover, a criterion function can return additional data that is stored in
+the log file (if logging is active). All of this is achieved by returning a dictionary
+instead of just a scalar float.
+
+For the simplest case, where only scalar optimizers are used, `criterion` returns a
+float. Here are two examples of this simple case.
+
+The **first example** represents `params` as a flat numpy array and returns a float.
+This would also be compatible with SciPy:
+
+```python
+def sphere(params: np.ndarray) -> float:
+    return params @ params
+```
+
+The **second example** also returns a float but uses a different format for the
+parameters:
+
+```python
+def dict_sphere(params: dict) -> float:
+    return params["a"] ** 2 + params["b"] ** 2
+```
+
+If the user wants the criterion function to be compatible with specialized optimizers
+for least-squares problems, the criterion function needs to return a dictionary. The
```

Comment:
I don't fully understand what the specialized optimizers require. Do I understand correctly that `"p_mean"` and `"p_std"` are not required, but just the least-squares residuals in `root_contributions`. Why does the criterion function needs to return a dictionary then?

Reply:
Thanks! I converted the example to two examples to make it clearer. But this partially just validates my feeling that our current way of specifying least-squares problems is very hard to describe.

Reply:
Thanks! Much clearer now 👍